src/test/java/org/broadinstitute/hellbender/utils/hmm/HMMUnitTest.java

```diff
@@ -0,0 +1,542 @@
+package org.broadinstitute.hellbender.utils.hmm;
+
+import org.apache.commons.math3.util.Pair;
+import org.broadinstitute.hellbender.utils.R.RScriptExecutor;
+import org.broadinstitute.hellbender.utils.test.BaseTest;
+import org.broadinstitute.hellbender.utils.tsv.DataLine;
+import org.broadinstitute.hellbender.utils.tsv.TableReader;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Unit test for HMM api classes including {@link ViterbiAlgorithm}, {@link ForwardBackwardAlgorithm} and
+ * {@link ForwardBackwardAlgorithm.Result}.
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public final class HMMUnitTest extends BaseTest {
```

Comment:
I'm uncomfortable about relying so heavily on R for these tests.  You should supplement it with a few HMMs that are a priori trivial.  For example:

* If priors and transitions are all equal, then every path must be equally likely.
* If one state has the likeliest prior and most likely transitions from any state are into that state, then the likeliest path is that state repeated over and over.
* If you have a long chain with (irreducible) transition matrix M, the state probabilities asymptotically approach the eigenvector with largest eigenvalue of M.  (I consider this test very important because it would be basically impossible to get the right answer by coincidence).